File: frontend/src/api/analysis.ts
import client from './client'

export interface ChatMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface SessionSummary {
  id: number
  player_name: string
  prompt_template_id: number | null
  title: string
  created_at: string
  updated_at: string
}

export interface SessionDetail extends SessionSummary {
  filter_opponent: string | null
  filter_deck: string | null
  filter_opponent_deck: string | null
  filter_format: string | null
  filter_date_from: string | null
  filter_date_to: string | null
  messages: Array<{
    id: number
    role: 'user' | 'assistant'
    content: string
    display_order: number
  }>
}

export interface PromptTemplate {
  id: number
  name: string
  content: string
  is_default: boolean
}

export interface QuestionItem {
  id: number
  text: string
  display_order: number
}

export interface QuestionSet {
  id: number
  name: string
  is_default: boolean
  items: QuestionItem[]
}

export interface StreamCallbacks {
  onDelta: (text: string) => void
  onDone: (sessionId: number) => void
  onError: (message: string) => void
}

export async function fetchSessions(player?: string): Promise<SessionSummary[]> {
  const res = await client.get<{ sessions: SessionSummary[] }>('/api/analysis/sessions', {
    params: player ? { player } : {},
  })
  return res.data.sessions
}

export async function fetchSessionDetail(id: number): Promise<SessionDetail> {
  const res = await client.get<SessionDetail>(`/api/analysis/sessions/${id}`)
  return res.data
}

export async function deleteSession(id: number): Promise<void> {
  await client.delete(`/api/analysis/sessions/${id}`)
}

export async function fetchPromptTemplates(): Promise<PromptTemplate[]> {
  const res = await client.get<{ templates: PromptTemplate[] }>('/api/prompt-templates')
  return res.data.templates
}

export async function fetchQuestionSets(): Promise<QuestionSet[]> {
  const res = await client.get<{ question_sets: QuestionSet[] }>('/api/question-sets')
  return res.data.question_sets
}

export async function streamChat(
  req: {
    player: string
    prompt_template_id?: number | null
    session_id?: number | null
    message: string
    history?: ChatMessage[]
    opponent?: string | null
    deck?: string | null
    opponent_deck?: string | null
    format?: string | null
    date_from?: string | null
    date_to?: string | null
  },
  callbacks: StreamCallbacks,
  signal?: AbortSignal,
): Promise<void> {
  let response: Response
  try {
    response = await fetch('http://localhost:8000/api/analysis/chat', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(req),
      signal,
    })
  } catch (e: any) {
    if (e?.name === 'AbortError') return
    callbacks.onError('サーバーへの接続に失敗しました')
    return
  }

  if (!response.ok) {
    const data = await response.json().catch(() => ({}))
    callbacks.onError(data.detail ?? `HTTP ${response.status}`)
    return
  }

  const reader = response.body!.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  let receivedDone = false
  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop()!

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        try {
          const event = JSON.parse(line.slice(6))
          if (event.delta !== undefined) {
            callbacks.onDelta(event.delta)
          } else if (event.done) {
            receivedDone = true
            callbacks.onDone(event.session_id)
          } else if (event.error) {
            receivedDone = true
            callbacks.onError(event.error)
          }
        } catch {}
      }
    }
  } catch (e: any) {
    if (e?.name !== 'AbortError') {
      callbacks.onError('ストリーミング中にエラーが発生しました')
    }
    return
  }
  if (!receivedDone) {
    callbacks.onError('接続が予期せず切断されました。バックエンドのログを確認してください。')
  }
}

File: backend/models/analysis.py
from datetime import datetime
from sqlalchemy import Integer, Text, DateTime, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship
from database import Base


class PromptTemplate(Base):
    __tablename__ = "prompt_templates"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    is_default: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    sessions: Mapped[list["AnalysisSession"]] = relationship(back_populates="prompt_template")


class QuestionSet(Base):
    __tablename__ = "question_sets"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    is_default: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    items: Mapped[list["QuestionItem"]] = relationship(
        back_populates="question_set",
        order_by="QuestionItem.display_order",
        cascade="all, delete-orphan",
    )


class QuestionItem(Base):
    __tablename__ = "question_items"
    __table_args__ = (
        Index("ix_question_items_question_set_id", "question_set_id"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    question_set_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("question_sets.id"), nullable=False
    )
    text: Mapped[str] = mapped_column(Text, nullable=False)
    display_order: Mapped[int] = mapped_column(Integer, nullable=False)

    question_set: Mapped["QuestionSet"] = relationship(back_populates="items")


class AnalysisSession(Base):
    __tablename__ = "analysis_sessions"
    __table_args__ = (
        Index("ix_analysis_sessions_player_name", "player_name"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    player_name: Mapped[str] = mapped_column(Text, nullable=False)
    prompt_template_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("prompt_templates.id"), nullable=True
    )
    title: Mapped[str] = mapped_column(Text, nullable=False)
    filter_opponent: Mapped[str | None] = mapped_column(Text, nullable=True)
    filter_deck: Mapped[str | None] = mapped_column(Text, nullable=True)
    filter_opponent_deck: Mapped[str | None] = mapped_column(Text, nullable=True)
    filter_format: Mapped[str | None] = mapped_column(Text, nullable=True)
    filter_date_from: Mapped[str | None] = mapped_column(Text, nullable=True)
    filter_date_to: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    prompt_template: Mapped["PromptTemplate | None"] = relationship(
        back_populates="sessions"
    )
    messages: Mapped[list["AnalysisMessage"]] = relationship(
        back_populates="session",
        order_by="AnalysisMessage.display_order",
        cascade="all, delete-orphan",
    )


class AnalysisMessage(Base):
    __tablename__ = "analysis_messages"
    __table_args__ = (
        Index("ix_analysis_messages_session_id", "session_id"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    session_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("analysis_sessions.id"), nullable=False
    )
    role: Mapped[str] = mapped_column(Text, nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    display_order: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    session: Mapped["AnalysisSession"] = relationship(back_populates="messages")
